src/line_parser.rs

```diff
@@ -0,0 +1,299 @@
+use wmidi::Note;
+
+use crate::{BeatNumber, Line, LineNote};
+
+#[derive(Debug)]
+enum State {
+    AwaitingNextToken,
+    ParsedNoteLetter { note_letter: char },
+    ParsedNoteName { note_name: String },
+}
+
+pub fn parse_line(line_str: &str) -> Line {
+    LineParser::new(line_str).parse()
+}
+
+struct LineParser {
+    chars: Vec<char>,
+    next_index: usize,
+    state: State,
+    currently_sustained_note: Option<(Note, BeatNumber)>,
+    current_beat_number: BeatNumber,
+}
+
+impl LineParser {
+    pub fn new(line_str: &str) -> Self {
+        Self {
+            chars: line_str.chars().collect(),
+            next_index: 0,
+            state: State::AwaitingNextToken,
```

Comment:
If you found it to be a readability improvement, you can `use State::*` in your imports, and then omit the `State::` prefix for the rest of the file.